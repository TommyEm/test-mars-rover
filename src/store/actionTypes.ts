export const UPDATE_ROVER_DIRECTION = 'UPDATE_ROVER_DIRECTION';
export const UPDATE_ROVER_POSITION = 'UPDATE_ROVER_POSITION';
export const UPDATE_COMMANDS = 'UPDATE_COMMANDS';
export const RESET = 'RESET';
