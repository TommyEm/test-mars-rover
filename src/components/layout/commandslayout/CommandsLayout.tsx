import React, { useCallback, useContext } from 'react';
import { COMMAND_LIST } from '../../../constants';
import { AppContext } from '../../../store/Store';
import * as actions from '../../../store/actionTypes';

import { StyledCommandsLayout } from './CommandsLayout.styled';
import { Label } from '../../form/label/Label';
import { Input } from '../../form/input/Input';
import { Button } from '../../button/Button';
import { moveRover } from '../../../lib/moveRover';
import { RoverCommands } from '../../../types/rover';


export interface ICommandsLayoutProps {
	className?: string;
}

export const CommandsLayout: React.FC<ICommandsLayoutProps> = ({
	className,
}: ICommandsLayoutProps) => {
	const { state: { commands, rover }, dispatch } = useContext(AppContext);

	const handleChangeInput = useCallback((command) => {
		if (COMMAND_LIST.includes(command.slice(-1))) { // The last key input is in the command list
			// Update the input
			dispatch({ type: actions.UPDATE_COMMANDS, payload: command });

			// Move the Rover
			const { newDir, newPos } = moveRover(command.slice(-1) as RoverCommands, rover.direction, rover.position);
			dispatch({ type: actions.UPDATE_ROVER_DIRECTION, payload: newDir });
			dispatch({ type: actions.UPDATE_ROVER_POSITION, payload: newPos });
		}
	}, [dispatch, rover.direction, rover.position]);

	const handleResetInput = useCallback(() => {
		dispatch({ type: actions.RESET });
	}, [dispatch]);

    return (
		<StyledCommandsLayout
			className={`CommandsLayout ${className}`}
			data-testid='test-CommandsLayout'
		>
			<Label forHtml='commands'>Enter some commands:</Label>
			<div className='CommandsLayout-input'>
				<Input
					name='commands'
					value={commands}
					onChange={e => handleChangeInput(e.currentTarget.value)}
				/>
				<Button onClick={handleResetInput}>Reset</Button>
			</div>

			<div className='CommandsLayout-help'>
				<h2>Commands list</h2>
				<dl className='CommandsLayout-command-list'>
					<dt className='CommandsLayout-command'>f</dt>
					<dd className='CommandsLayout-desc'>Move forward</dd>
					<dt className='CommandsLayout-command'>l</dt>
					<dd className='CommandsLayout-desc'>Rotate left</dd>
					<dt className='CommandsLayout-command'>r</dt>
					<dd className='CommandsLayout-desc'>Rotate right</dd>
				</dl>
			</div>
		</StyledCommandsLayout>
	);
}
