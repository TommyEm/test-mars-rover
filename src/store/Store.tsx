import React, { createContext, useReducer } from 'react';
import { reducer } from './reducer';
import { IAppContext, IAppState } from '../types/store';
import * as constants from '../constants';



export const initialState: IAppState = {
	commands: '',
	rover: {
		direction: constants.ROVER_DEFAULT_DIRECTION,
		position: constants.ROVER_DEFAULT_POSITION,
	},
};


export const AppContext = createContext<IAppContext>({
	state: initialState,
	dispatch: () => {},
});

interface IStore {
	children: React.ReactNode;
}

export const Store = ({ children }: IStore) => {
	const [state, dispatch] = useReducer(reducer, initialState);

	return (
		<AppContext.Provider value={{ state, dispatch }}>
			{children}
		</AppContext.Provider>
	);
};
