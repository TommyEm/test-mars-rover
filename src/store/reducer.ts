import produce from 'immer';
import { IAction, IAppState } from "../types/store";
import { initialState } from "./Store";
import * as actions from '../store/actionTypes';


export const reducer = produce((draft: IAppState, action: IAction) => {
	switch (action.type) {
		case actions.UPDATE_ROVER_DIRECTION:
			draft.rover.direction = action.payload;
			break;

		case actions.UPDATE_ROVER_POSITION:
			draft.rover.position = action.payload;
			break;

		case actions.UPDATE_COMMANDS:
			draft.commands = action.payload;
			break;

		case actions.RESET:
			draft.commands = '';
			draft.rover = { ...initialState.rover };
			break;
	}
});
