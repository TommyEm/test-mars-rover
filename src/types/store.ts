import { Dispatch, ReducerAction } from 'react';
import { reducer } from '../store/reducer';
import { RoverDirection, RoverPosition } from '../types/rover';


export interface IAppState {
	commands: string;
	rover: {
		direction: RoverDirection;
		position: RoverPosition;
	},
}

export interface IAppContext {
	state: IAppState;
	dispatch: Dispatch<ReducerAction<typeof reducer>>;
}

export interface IAction {
	type: string;
	payload: any;
}
