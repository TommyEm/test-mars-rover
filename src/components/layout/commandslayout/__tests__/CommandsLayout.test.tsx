import React from 'react';
import {
	cleanup,
	// fireEvent,
	render,
	screen
} from '@testing-library/react';
import { composeStories } from '@storybook/testing-react';

import * as stories from '../CommandsLayout.stories';
import { Store } from '../../../../store/Store';

const { Default } = composeStories(stories);


afterEach(cleanup);
const storyComponent = <Store><Default /></Store>;

it('should render a commands layout', () => {
	render(storyComponent);
	expect(screen.getByTestId('test-CommandsLayout')).toBeVisible();
});

it('matches snapshot', () => {
	const tree = render(storyComponent);
	expect(tree).toMatchSnapshot();
});

// it('should update the store', () => {
// 	const onChange = jest.fn();
// 	render(storyComponent);
// 	fireEvent.change(screen.getByTestId('test-Input'), { target: { value: 'f' } });
// });
