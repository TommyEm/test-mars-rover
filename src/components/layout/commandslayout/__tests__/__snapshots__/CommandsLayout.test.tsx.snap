// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-bdfBQB goExrx CommandsLayout undefined"
        data-testid="test-CommandsLayout"
      >
        <label
          class="sc-gsTEea jgTHUl Label undefined"
          data-testid="test-Label"
        >
          Enter some commands:
        </label>
        <div
          class="CommandsLayout-input"
        >
          <input
            class="sc-dlfnuX kVhtmp Input undefined"
            data-testid="test-Input"
            id="commands"
            name="commands"
            type="text"
            value=""
          />
          <button
            class="sc-hKgJUU kGjTEm Button"
            data-testid="test-Button"
          >
            Reset
          </button>
        </div>
        <div
          class="CommandsLayout-help"
        >
          <h2>
            Commands list
          </h2>
          <dl
            class="CommandsLayout-command-list"
          >
            <dt
              class="CommandsLayout-command"
            >
              f
            </dt>
            <dd
              class="CommandsLayout-desc"
            >
              Move forward
            </dd>
            <dt
              class="CommandsLayout-command"
            >
              l
            </dt>
            <dd
              class="CommandsLayout-desc"
            >
              Rotate left
            </dd>
            <dt
              class="CommandsLayout-command"
            >
              r
            </dt>
            <dd
              class="CommandsLayout-desc"
            >
              Rotate right
            </dd>
          </dl>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-bdfBQB goExrx CommandsLayout undefined"
      data-testid="test-CommandsLayout"
    >
      <label
        class="sc-gsTEea jgTHUl Label undefined"
        data-testid="test-Label"
      >
        Enter some commands:
      </label>
      <div
        class="CommandsLayout-input"
      >
        <input
          class="sc-dlfnuX kVhtmp Input undefined"
          data-testid="test-Input"
          id="commands"
          name="commands"
          type="text"
          value=""
        />
        <button
          class="sc-hKgJUU kGjTEm Button"
          data-testid="test-Button"
        >
          Reset
        </button>
      </div>
      <div
        class="CommandsLayout-help"
      >
        <h2>
          Commands list
        </h2>
        <dl
          class="CommandsLayout-command-list"
        >
          <dt
            class="CommandsLayout-command"
          >
            f
          </dt>
          <dd
            class="CommandsLayout-desc"
          >
            Move forward
          </dd>
          <dt
            class="CommandsLayout-command"
          >
            l
          </dt>
          <dd
            class="CommandsLayout-desc"
          >
            Rotate left
          </dd>
          <dt
            class="CommandsLayout-command"
          >
            r
          </dt>
          <dd
            class="CommandsLayout-desc"
          >
            Rotate right
          </dd>
        </dl>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
